firewood/src/storage/linear/proposed.rs

```diff
@@ -4,42 +4,40 @@
 use std::collections::BTreeMap;
 use std::fmt::Debug;
 use std::io::{Error, Read};
-use std::marker::PhantomData;
 use std::sync::Arc;
 
-use super::{LinearStore, ReadLinearStore, WriteLinearStore};
+use super::layered::{Layer, LayeredReader};
+use super::{ReadLinearStore, WriteLinearStore};
 
 /// [Proposed] is a [LinearStore] state that contains a copy of the old and new data.
-/// The P type parameter indicates the state of the linear store for it's parent,
-/// which could be a another [Proposed] or is [FileBacked](super::filebacked::FileBacked)
-/// The M type parameter indicates the mutability of the proposal, either read-write or readonly
 #[derive(Debug)]
-pub(crate) struct Proposed<P: ReadLinearStore, M> {
-    pub(crate) new: BTreeMap<u64, Box<[u8]>>,
-    old: BTreeMap<u64, Box<[u8]>>,
-    pub(crate) parent: Arc<LinearStore<P>>,
-    phantom: PhantomData<M>,
+pub(crate) struct Proposed {
```

Comment:
I don't like removing M here. This means all `Proposed` can be mutated by mistake. Lets avoid this problem by having two types:
`ProposedMutable` and `ProposedImmutable` which the compiler can do for you with `Proposed<Mutable>` and `Proposed<Immutable>` and some `PhantomData`. Then, we only implement `WriteLinearStore` on `Proposed<Mutable>`.

Reply:
I removed this based on @richardpringle 's [comment](https://github.com/ava-labs/firewood/pull/633#discussion_r1569619102) which I think makes a reasonable point. If someone has a `&Proposed`, then they can't mutate it, but if they have a `&mut proposed` then they can. If they have a `dyn ReadLinearStore` they can't mutate it, but if they have a `dyn WriteLinearStore` they can. 

What does having these types `Proposed<Mutable>` and `Proposed<Immutable>` buy us that `mut` and the `dyn`s don't?

Reply:
It buys us that you can't call `write` on an immutable proposal, and this is caught at compile time.